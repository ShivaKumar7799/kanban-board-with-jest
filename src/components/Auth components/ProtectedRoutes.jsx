import React, { useEffect } from "react";
import NavBar from "../Reusable components/NavBar"
import { useNavigate } from "react-router-dom";
import { useDispatch } from "react-redux";
import { signInUser } from "../../redux/features/userSlice";
import { saveTasks } from "../../redux/features/tasksSlice";
import apiService from "../../Services/Api Services/ApiServices";

function ProtectedRoutes({ children }) {
  const dispatch = useDispatch();
  const navigate = useNavigate();

  useEffect(() => {
    const user = JSON.parse(localStorage.getItem("user"));

    const getTasks = () => {
      apiService
        .get(`/tasks`)
        .then((res) => {
          const updatedTasks = res.data.filter(
            (task) => task.email === user?.email
          );
          dispatch(saveTasks(updatedTasks));
        })
        .catch((err) => console.log(err));
    };

    if (user === null) {
      navigate("/");
      return;
    } else {
      getTasks();
      dispatch(signInUser(user));
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  return (
    <div>
      <NavBar />
      {children}
    </div>
  );
}

export default ProtectedRoutes;
