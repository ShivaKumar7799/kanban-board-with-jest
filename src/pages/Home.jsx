import React, { useEffect, useState } from "react";
import { useDispatch } from "react-redux";
import { signInUser } from "../redux/features/userSlice";
import { useNavigate, Link } from "react-router-dom";
import { useFormik } from "formik";
import apiService from "../Services/Api Services/ApiServices";

function Home() {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const [showPassword, setShowPassword] = useState(false);

  const [users, setUsers] = useState([]);
  const [invalidUser, setInvalidUser] = useState(false);

  useEffect(() => {
    apiService.get(`/signUp`).then((res) => setUsers(res.data)).catch(err => console.log("signup api error"));
  }, []);

  const loginValidation = (values) => {
    setInvalidUser(false);
    const errors = {};

    if (values.userNameOrEmail === "") {
      errors.userNameOrEmail = "UserName/email is Mandatory";
    }

    if (values.password === "") {
      errors.password = "Password is Required";
    } else if (values.password.length < 8) {
      errors.password = "Password should contain minimun 8 characters";
    } else if (values.password.length > 20) {
      errors.password = "password shouldn't contain more than 20 characters";
    }

    return errors;
  };

  const loginFormik = useFormik({
    initialValues: {
      userNameOrEmail: "",
      password: "",
    },
    validate: loginValidation,
    onSubmit: (values) => {
      const validUser = users.filter(
        (user) =>
          (user.userName === values.userNameOrEmail ||
            user.email === values.userNameOrEmail) &&
          user.password === values.password
      );
      if (validUser.length > 0) {
        navigate("/dashBoard");
        dispatch(signInUser(validUser[0]));
        localStorage.setItem("user", JSON.stringify(validUser[0]));
      } else {
        setInvalidUser(true);
      }
    },
  });
  return (
    <div className="d-flex justify-content-center min-vh-100 bg-dark align-items-center">
      <div
        className="w-50 border border-primary rounded py-3 px-5 bg-white h-75"
        style={{ minHeight: "375px" }}
      >
        <h1 className="text-center text-dark"> Kanban Board</h1>

        <form onSubmit={loginFormik.handleSubmit}>
          <div className="d-flex flex-column">
            <label htmlFor="userNameOrEmail">Username/email</label>
            <input
              type="text"
              name="userNameOrEmail"
              id="userNameOrEmail"
              className="form-control"
              placeholder="Please enter your username or email here"
              value={loginFormik.userNameOrEmail}
              onChange={loginFormik.handleChange}
              onBlur={loginFormik.handleBlur}
            />
            {loginFormik.errors.userNameOrEmail &&
              loginFormik.touched.userNameOrEmail && (
                <span className="text-danger py-2">
                  {loginFormik.errors.userNameOrEmail}
                </span>
              )}
          </div>

          <div className="d-flex flex-column">
            <label htmlFor="password">password*</label>
            <input
              type={`${showPassword ? "text" : "password"}`}
              value={loginFormik.password}
              className="form-control"
              name="password"
              id="password"
              placeholder="Enter your password"
              onChange={loginFormik.handleChange}
              onBlur={loginFormik.handleBlur}
            />
            <span
              className="text-primary cursor-pointer"
              onClick={() => setShowPassword(!showPassword)}
            >
              {showPassword ? "hide password" : "show password"}{" "}
            </span>
            {loginFormik.errors.password && loginFormik.touched.password && (
              <span className="text-danger py-2">
                {loginFormik.errors.password}
              </span>
            )}
          </div>

          <div>
            {invalidUser && (
              <div className="text-danger pt-2">
                Invalid Username or Password
              </div>
            )}
          </div>

          <div className="d-flex justify-content-center py-3">
            <input
              className="btn btn-success btn-lg"
              type="submit"
              value="Login"
            />
          </div>

          <div>
            <span>
              Don't have account? <Link to="/signUp">Register Here</Link>
            </span>
          </div>
        </form>
      </div>
    </div>
  );
}

export default Home;
