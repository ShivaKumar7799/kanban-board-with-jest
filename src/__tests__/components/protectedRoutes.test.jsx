import React from 'react';
import { render,screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store'; // Import your Redux store configuration
import ProtectedRoutes from '../../components/Auth components/ProtectedRoutes';
import Home from '../../pages/Home'

const mockStore = configureStore([]);

describe('ProtectedRoutes', () => {

  test('snapshot of loader component', () => {
    const store = mockStore({ user: { email: 'test@example.com' } });
    const { container } = render(<Provider store={store}>
      <MemoryRouter>
        <ProtectedRoutes>
          <Home />
        </ProtectedRoutes>
      </MemoryRouter>
    </Provider>);
    expect(container).toMatchSnapshot();
  });

  test('renders children components when user is authenticated', async() => {
    const store = mockStore({ user: { email: 'shiva@gmail.com' } });
    render(
      <Provider store={store}>
        <MemoryRouter>
          <ProtectedRoutes>
            <Home />
          </ProtectedRoutes>
        </MemoryRouter>
      </Provider>
    );

    const childComponent = await screen.findByText(/kanban board/i);
    expect(childComponent).toBeInTheDocument();
  });
});
