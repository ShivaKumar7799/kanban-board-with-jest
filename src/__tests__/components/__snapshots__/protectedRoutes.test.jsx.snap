// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProtectedRoutes snapshot of loader component 1`] = `
<div>
  <div>
    <div
      class="bg-dark d-flex py-3 justify-content-between align-items-center"
    >
      <div
        class="d-flex flex-row"
      >
        <div
          class="mx-3"
        >
          <a
            class="text-decoration-none "
            href="/dashBoard"
          >
            <span
              class="text-white "
            >
              DashBoard
            </span>
          </a>
        </div>
        <div>
          <a
            class="text-decoration-none "
            href="/taskManagement"
          >
            <span
              class="text-white "
            >
              Task Management
            </span>
          </a>
        </div>
      </div>
      <div>
        <button
          class="btn btn-primary mx-3"
        >
          Logout
        </button>
      </div>
    </div>
    <div
      class="d-flex justify-content-center min-vh-100 bg-dark align-items-center"
    >
      <div
        class="w-50 border border-primary rounded py-3 px-5 bg-white h-75"
        style="min-height: 375px;"
      >
        <h1
          class="text-center text-dark"
        >
           Kanban Board
        </h1>
        <form>
          <div
            class="d-flex flex-column"
          >
            <label
              for="userNameOrEmail"
            >
              Username/email
            </label>
            <input
              class="form-control"
              id="userNameOrEmail"
              name="userNameOrEmail"
              placeholder="Please enter your username or email here"
              type="text"
              value=""
            />
          </div>
          <div
            class="d-flex flex-column"
          >
            <label
              for="password"
            >
              password*
            </label>
            <input
              class="form-control"
              id="password"
              name="password"
              placeholder="Enter your password"
              type="password"
              value=""
            />
            <span
              class="text-primary cursor-pointer"
            >
              show password
               
            </span>
          </div>
          <div />
          <div
            class="d-flex justify-content-center py-3"
          >
            <input
              class="btn btn-success btn-lg"
              type="submit"
              value="Login"
            />
          </div>
          <div>
            <span>
              Don't have account? 
              <a
                href="/signUp"
              >
                Register Here
              </a>
            </span>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
